### PR TITLE
Replace int with std::chrono in for the timer->setInterval() argument

### PR DESCRIPTION
**Pull request by shaavan**
The PR is a follow-up to #517

- It addresses the change suggested in [this](https://github.com/bitcoin-core/gui/pull/517#pullrequestreview-850260826) comment.
- This PR changes the type of `msecsPerSample` from **int** to **std::chrono::minutes** and makes other relevant subsequent changes that were limited to the **trafficgraphwidget** file.

### CODE REVIEW

File: src/qt/trafficgraphwidget.h [hunk 1]
```diff
@@ -22,7 +24,7 @@ class TrafficGraphWidget : public QWidget
 public:
     explicit TrafficGraphWidget(QWidget *parent = nullptr);
     void setClientModel(ClientModel *model);
-    int getGraphRangeMins() const;
+    std::chrono::minutes getGraphRangeMins() const;
```
Comment (hebasto) [inline src/qt/trafficgraphwidget.h:27]:
```suggestion
    std::chrono::minutes getGraphRange() const;
```

Also s/`setGraphRangeMins`/`setGraphRange`/

File: src/qt/trafficgraphwidget.h [hunk 2]
```diff
@@ -37,7 +39,7 @@ public Q_SLOTS:
 
     QTimer *timer;
     float fMax;
-    int nMins;
+    std::chrono::minutes nMins;
```
Comment (hebasto) [inline src/qt/trafficgraphwidget.h:42]:
```suggestion
    std::chrono::minutes m_range;
```

File: src/qt/trafficgraphwidget.cpp [hunk 3]
```diff
@@ -11,18 +11,21 @@
...
 #define YMARGIN                 10
 
+using namespace std::chrono_literals;
+
 TrafficGraphWidget::TrafficGraphWidget(QWidget *parent) :
     QWidget(parent),
     timer(nullptr),
     fMax(0.0f),
-    nMins(0),
+    nMins(0min),
```
Comment (hebasto) [inline src/qt/trafficgraphwidget.cpp:28]: Move into a member initializer?

File: src/qt/trafficgraphwidget.h (hunk 1 above)
Comment (shaavan) [reply src/qt/trafficgraphwidget.h]: Addressed in 9bb48bf657c7f2d36309807fbc7b55d0ab581a76

File: src/qt/trafficgraphwidget.h (hunk 2 above)
Comment (shaavan) [reply src/qt/trafficgraphwidget.h]: Addressed in 9bb48bf657c7f2d36309807fbc7b55d0ab581a76

File: src/qt/trafficgraphwidget.cpp (hunk 3 above)
Comment (shaavan) [reply src/qt/trafficgraphwidget.cpp]: Addressed in 9bb48bf657c7f2d36309807fbc7b55d0ab581a76

**Review by jonatack — COMMENTED** [review]
Concept ACK 9bb48bf657c7f2d36309807fbc7b55d0ab581a76

**Comment by RandyMcMillan** [comment]
nACK 9bb48bf657c7f2d36309807fbc7b55d0ab581a76

This is breaking the Network Traffic tab on MacOS X86_64 &  arm64

---
Darwin ₿ 19.6.0 Darwin Kernel Version 19.6.0: Tue Oct 12 18:34:05 PDT 2021; root:xnu-6153.141.43~1/RELEASE_X86_64 x86_64

Screen Shot 2022-01-14 at 12 17 27 AM
---
Darwin ₿ 20.5.0 Darwin Kernel Version 20.5.0: Sat May  8 05:10:31 PDT 2021; root:xnu-7195.121.3~9/RELEASE_ARM64_T8101 arm64

Screen Shot 2022-01-14 at 12 17 34 AM
---

**Review by hebasto — COMMENTED** [review]
Tested 9bb48bf657c7f2d36309807fbc7b55d0ab581a76 on Linux Mint 20.3 (Qt 5.12.8) -- the "Network Traffic" tab shows a black window.

File: src/qt/trafficgraphwidget.cpp [hunk 4]
```diff
@@ -153,10 +155,10 @@ void TrafficGraphWidget::updateRates()
     update();
 }
 
-void TrafficGraphWidget::setGraphRangeMins(int mins)
+void TrafficGraphWidget::setGraphRange(std::chrono::minutes mins)
 {
-    nMins = mins;
-    int msecsPerSample = nMins * 60 * 1000 / DESIRED_SAMPLES;
+    m_range = std::chrono::minutes(mins);
+    auto msecsPerSample = m_range * 60 * 1000 / DESIRED_SAMPLES;
```
Comment (hebasto) [inline src/qt/trafficgraphwidget.cpp:161]: https://github.com/bitcoin-core/gui/pull/524#pullrequestreview-853590647
> Tested [9bb48bf](https://github.com/bitcoin-core/gui/commit/9bb48bf657c7f2d36309807fbc7b55d0ab581a76) on Linux Mint 20.3 (Qt 5.12.8) -- the "Network Traffic" tab shows a black window.

Could be fixed with
```suggestion
    const auto msecs_per_sample{std::chrono::duration_cast<std::chrono::milliseconds>(m_range) / DESIRED_SAMPLES};
```

**Comment by shaavan** [comment]
Updated from ([pr524.02](https://github.com/shaavan/gui/tree/pr524.02) -> [pr524.03](https://github.com/shaavan/gui/tree/pr524.03))
Addressed @jonatack and @hebasto’s suggestions.

- Done proper casting of **m_range** variable from **mins** to **milliseconds**.
- Used `duration_casting` instead of `static_casting`, as suggested by @hebasto. The reason for doing so is explained in [this](https://stackoverflow.com/a/44277832) article.
- Incorporated other **nit** suggestions.

Tested that the network traffic graph is working correctly now.

File: src/qt/trafficgraphwidget.cpp (hunk 4 above)
Comment (shaavan) [reply src/qt/trafficgraphwidget.cpp]: Thanks, @hebasto for catching it and giving suggestions on how to correct this misbehavior

Addressed in 5e6ad435b0ed634fca5745668bf949799b306bff

File: src/qt/rpcconsole.cpp [hunk 5]
```diff
@@ -1140,7 +1140,7 @@ void RPCConsole::on_sldGraphRange_valueChanged(int value)
 
 void RPCConsole::setTrafficGraphRange(int mins)
 {
-    ui->trafficGraph->setGraphRangeMins(mins);
+    ui->trafficGraph->setGraphRange(std::chrono::minutes(mins));
```
[the remark this answers is not in the document]
Comment (shaavan) [reply src/qt/rpcconsole.cpp]: Thanks, @jonatack, for nit suggestions, as well as catching the missing header. I intended to add the header, but for some reason, I missed adding it before updating the PR for some reason.

Addressed in 5e6ad43

File: src/qt/trafficgraphwidget.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (shaavan) [reply src/qt/trafficgraphwidget.cpp]: Addressed in 5e6ad43

File: src/qt/trafficgraphwidget.cpp [hunk 6]
```diff
@@ -42,9 +44,9 @@ void TrafficGraphWidget::setClientModel(ClientModel *model)
     }
 }
 
-int TrafficGraphWidget::getGraphRangeMins() const
+std::chrono::minutes TrafficGraphWidget::getGraphRange() const
 {
-    return nMins;
+    return m_range;
```
[the remark this answers is not in the document]
Comment (shaavan) [reply src/qt/trafficgraphwidget.cpp]: I think it's better to incorporate the correct way of writing the code from the get-go.

Addressed in 5e6ad43

File: src/qt/trafficgraphwidget.cpp [hunk 7]
```diff
@@ -11,18 +11,20 @@
...
+#include <chrono>
 #include <cmath>
 
 #define DESIRED_SAMPLES         800
 
 #define XMARGIN                 10
 #define YMARGIN                 10
 
+using namespace std::chrono_literals;
+
```
Comment (hebasto) [inline src/qt/trafficgraphwidget.cpp:23]: Why this change?

File: src/qt/trafficgraphwidget.cpp [hunk 8]
```diff
@@ -153,12 +152,12 @@ void TrafficGraphWidget::updateRates()
     update();
 }
 
-void TrafficGraphWidget::setGraphRangeMins(int mins)
+void TrafficGraphWidget::setGraphRange(std::chrono::minutes mins)
 {
-    nMins = mins;
-    int msecsPerSample = nMins * 60 * 1000 / DESIRED_SAMPLES;
+    m_range = std::chrono::minutes{mins};
```
Comment (hebasto) [inline src/qt/trafficgraphwidget.cpp:157]: Maybe just
```suggestion
    m_range = mins;
```
?

**Review by hebasto — COMMENTED** [review]
Approach ACK 5e6ad435b0ed634fca5745668bf949799b306bff, tested on Linux Mint 20.3 (Qt 5.12.8).

About commit message:
- please use "qt: " prefix (see https://github.com/bitcoin-core/gui/blob/master/CONTRIBUTING.md#creating-the-pull-request)
- what about "qt, refactor: Use std::chrono in TrafficGraphWidget class"?
- references like `PR#517` do not work well with separated repositories that use the monotree; you could use `bitcoin-core/gui#517`, or just drop it

File: src/qt/trafficgraphwidget.h [hunk 9]
```diff
@@ -22,22 +24,22 @@ class TrafficGraphWidget : public QWidget
...
-    int getGraphRangeMins() const;
+    std::chrono::minutes getGraphRange() const;
 
 protected:
     void paintEvent(QPaintEvent *) override;
 
 public Q_SLOTS:
     void updateRates();
-    void setGraphRangeMins(int mins);
+    void setGraphRange(std::chrono::minutes mins);
```
Comment (hebasto) [inline src/qt/trafficgraphwidget.h:34]: Maybe
```suggestion
    void setGraphRange(std::chrono::minutes new_range);
```
?

**Comment by shaavan** [comment]
Updated from  5e6ad43 to f7a19ef ([pr524.03](https://github.com/shaavan/gui/tree/pr524.03) -> [pr524.04](https://github.com/shaavan/gui/tree/pr524.04))
Addressed @hebasto suggestions

Changes:
- Removed the redundant `using namespace std::chrono_literals;` statement.
- Removed unnecessary type conversion of `chrono::minutes` variable.
- Renamed `mins` -> `new_range`.
- Made appropriate changes to commit message as suggested [here](https://github.com/bitcoin-core/gui/pull/524#pullrequestreview-853649608).

File: src/qt/trafficgraphwidget.cpp (hunk 7 above)
Comment (shaavan) [reply src/qt/trafficgraphwidget.cpp]: I checked through the file and made sure that this statement is not needed.

Removed in f7a19ef774ef92ce348215593e3590a750c345e1

File: src/qt/trafficgraphwidget.cpp (hunk 8 above)
Comment (shaavan) [reply src/qt/trafficgraphwidget.cpp]: The type-conversion here seems redundant.

Addressed in f7a19ef774ef92ce348215593e3590a750c345e1

File: src/qt/trafficgraphwidget.h (hunk 9 above)
Comment (shaavan) [reply src/qt/trafficgraphwidget.h]: I think `new_range` is a more suitable name for the argument, and now its unit is cleared by its type (i.e., minutes)

Addressed in f7a19ef774ef92ce348215593e3590a750c345e1

**Review by hebasto — APPROVED** [review]
ACK f7a19ef774ef92ce348215593e3590a750c345e1

**Comment by RandyMcMillan** [comment]
tACK f7a19ef774ef92ce348215593e3590a750c345e1

on MacOS x86_64 and Arm64

**Review by Gemk83 — APPROVED** [review]

File: src/qt/trafficgraphwidget.h [hunk 10]
```diff
@@ -22,22 +24,22 @@ class TrafficGraphWidget : public QWidget
 public:
     explicit TrafficGraphWidget(QWidget *parent = nullptr);
     void setClientModel(ClientModel *model);
-    int getGraphRangeMins() const;
+    std::chrono::minutes getGraphRange() const;
```
Comment (promag) [inline src/qt/trafficgraphwidget.h:27]: Just remove? It's not used.

**Review by promag — COMMENTED** [review]
Code review ACK f7a19ef774ef92ce348215593e3590a750c345e1.

File: src/qt/trafficgraphwidget.h (hunk 10 above)
Comment (RandyMcMillan) [reply src/qt/trafficgraphwidget.h:27]: @promag @shaavan 

I concur - It appears to be part of the original implementation but never used for anything...

it can and should be removed...

https://github.com/bitcoin-core/gui/blame/16781e1bc9f8ffc721ebea73434e0066957bc959/src/qt/trafficgraphwidget.cpp#L45


Screen Shot 2022-01-20 at 1 16 34 PM

Comment (RandyMcMillan) [reply src/qt/trafficgraphwidget.h:27]: @promag - great catch!

**Comment by shaavan** [comment]
Updated from ([pr524.04](https://github.com/shaavan/gui/tree/pr524.04) -> [pr524.05](https://github.com/shaavan/gui/tree/pr524.05))
Addressed @promag suggestions

Changes:
- Removed the unused function `getGraphRange()` from the **TrafficGraphWidget** class.

File: src/qt/trafficgraphwidget.h (hunk 10 above)
Comment (shaavan) [reply src/qt/trafficgraphwidget.h]: Thanks @promag for catching this. Addressed in 5efd391eb1ea1a3551b2335827e4b5577be2aca3

**Review by w0xlt — APPROVED** [review]
tACK 5efd391 on Ubuntu 21.10 (Qt 5.15.2).

**Review by hebasto — APPROVED** [review]
re-ACK 5efd391eb1ea1a3551b2335827e4b5577be2aca3

File: src/qt/trafficgraphwidget.cpp [hunk 11]
```diff
@@ -153,12 +148,12 @@ void TrafficGraphWidget::updateRates()
     update();
 }
 
-void TrafficGraphWidget::setGraphRangeMins(int mins)
+void TrafficGraphWidget::setGraphRange(std::chrono::minutes new_range)
```
Comment (promag) [inline src/qt/trafficgraphwidget.cpp:151]: Drop `m_range` and rename this arg to `range`.

Comment (promag) [reply src/qt/trafficgraphwidget.cpp:151]: Unless you early return like `if (m_range == new_range) return`, but I think this won't happen in practice.

Comment (hebasto) [reply src/qt/trafficgraphwidget.cpp:151]: > Drop `m_range` and rename this arg to `range`.

In a separate (first?) commit, no?

Comment (promag) [reply src/qt/trafficgraphwidget.cpp:151]: Why not the same commit? Just drop `nMin` instead of rename.

Comment (hebasto) [reply src/qt/trafficgraphwidget.cpp:151]: (a) getting rid of `TrafficGraphWidget::getGraphRangeMins()` and `TrafficGraphWidget::m_range`, and (b) changing type of `TrafficGraphWidget::setGraphRange()` parameter looks really as two focused commits, no?

Comment (promag) [reply src/qt/trafficgraphwidget.cpp:151]: Either way loos fine to me 👍

**Comment by shaavan** [comment]
Updated from 5efd391 to bdf7cb9 ([pr524.05](https://github.com/shaavan/gui/tree/pr524.05) -> [pr524.06](https://github.com/shaavan/gui/tree/pr524.06))
Addressed @promag suggestion

Changes:
- Removed `m_range` as a member variable of the class TrafficGraphWidget and replaced it with a local `range` variable.

File: src/qt/trafficgraphwidget.cpp (hunk 11 above)
Comment (shaavan) [reply src/qt/trafficgraphwidget.cpp:151]: I think this is a valid suggestion. Better to get rid of a redundant member variable.

Addressed in bdf7cb9f0c17bdeeb8389d4b9140aca65e08ac71

@hebasto
> (a) getting rid of TrafficGraphWidget::getGraphRangeMins() and TrafficGraphWidget::nMin, and (b) refactoring of TrafficGraphWidget::setGraphRangeMins() (renaming and parameter type changing) looks really as two focused commits, no?

I agree with your reasoning. However, (a) and (b) are not mutually exclusive, and hence I was having a hard time splitting these two changes while keeping the commits meaningful. Therefore, I reckoned it would be best to make these changes in one commit.

**Comment by RandyMcMillan** [comment]
Approach ACK
Removing the TrafficGraphWidget::getGraphRangeMins() function nACK

I use int TrafficGraphWidget::getGraphRangeMins() const in PR #539 (which fixes issue #532)

I have added a variation to of this PR to #539 https://github.com/bitcoin-core/gui/pull/539/commits/c25d3f42084926c92ae2051278cd785cb1907d56 to ensure compatibility with the chrono lib.

https://github.com/bitcoin-core/gui/blob/c25d3f42084926c92ae2051278cd785cb1907d56/src/qt/trafficgraphwidget.cpp#L46

### Note

https://github.com/bitcoin-core/gui/pull/524/files#diff-da38549cd85e0c422891b8ef1eb3cc33e76f0b80d15ca63aa7ad6f065c8ba694L44

File: src/qt/trafficgraphwidget.h (hunk 10 above)
Comment (RandyMcMillan) [reply src/qt/trafficgraphwidget.h]: > Approach ACK Removing the TrafficGraphWidget::getGraphRangeMins() function nACK
> 
> I use int TrafficGraphWidget::getGraphRangeMins() const in PR #539 (which fixes issue #532)
> 
> I have added a variation to of this PR to #539 [c25d3f4](https://github.com/bitcoin-core/gui/commit/c25d3f42084926c92ae2051278cd785cb1907d56) to ensure compatibility with the chrono lib.
> 
> https://github.com/bitcoin-core/gui/blob/c25d3f42084926c92ae2051278cd785cb1907d56/src/qt/trafficgraphwidget.cpp#L46
> 
> ### Note
> https://github.com/bitcoin-core/gui/pull/524/files#diff-da38549cd85e0c422891b8ef1eb3cc33e76f0b80d15ca63aa7ad6f065c8ba694L44

**Comment by shaavan** [comment]
> I use int TrafficGraphWidget::getGraphRangeMins() const in PR #539 (which fixes issue #532)

I looked through #532, but I couldn't find any use case of `getGraphRangeMins()` there. So I don't think there should be an issue removing this function in this PR.

> I have added a variation of this PR to #539 c25d3f4 to ensure compatibility with the chrono lib.

I looked through it, and it looked good. However, I reckon the primary purpose of #539 is to fix an issue by adding a scaling feature. And the purpose of this PR is refactoring. So I think it would be best to keep these PRs separate.

**Comment by RandyMcMillan** [comment]
If you insist - but to scale the graph - we will have to know the current value of the slider - otherwise the graph can't calculate how many samples to display. It will have to be re-added - it is easier to leave it in. 

https://github.com/bitcoin-core/gui/pull/539

**Comment by shaavan** [comment]
> It will have to be re-added - it is easier to leave it in.

I think you are right. Let me keep the `GetGraphRange()` function and make it consistent with the previous changes.

**Comment by shaavan** [comment]
Updated from bdf7cb9 to cad0e0a ([pr524.06](https://github.com/shaavan/gui/tree/pr524.06) -> [pr524.07](https://github.com/shaavan/gui/tree/pr524.07))
Addressed @RandyMcMillan comment

Changes:
- Restored the `getGraphRange()` function, as it is used in #539. So removing and re-adding the function doesn’t seem optimal.

The updated PR correctly compiles on Ubuntu 20.04, with the Network Traffic widget showing the correct graph.

**Comment by shaavan** [comment]
Reverted back to f7a19ef774ef92ce348215593e3590a750c345e1 ([pr524.07](https://github.com/shaavan/gui/tree/pr524.07) -> [pr524.04](https://github.com/shaavan/gui/tree/pr524.04), [diff](https://github.com/shaavan/gui/compare/pr524.07..pr524.04))

Reason:
- The is almost no difference between pr524.07 and pr524.04 (see the [diff](https://github.com/shaavan/gui/compare/pr524.07..pr524.04)).
- f7a19ef774ef92ce348215593e3590a750c345e1 had quite a few ACKs behind it, by:
    - @hebasto: https://github.com/bitcoin-core/gui/pull/524#pullrequestreview-855595100
    - @RandyMcMillan: https://github.com/bitcoin-core/gui/pull/524#issuecomment-1015866815
    - @promag: https://github.com/bitcoin-core/gui/pull/524#pullrequestreview-858455886

P.S.: To avoid any unintended changes between the PR and f7a19ef774ef92ce348215593e3590a750c345e1, I used the following command to update the PR:
```
git reset --hard HEAD^
git cherry-pick f7a19ef774ef92ce348215593e3590a750c345e1
```
The commit id has changed (which I was not able to restore) but the changes are identical f7a19ef774ef92ce348215593e3590a750c345e1

**Comment by shaavan** [comment]
Restored PR to previous commit f7a19ef774ef92ce348215593e3590a750c345e1. The reason for doing so is explained here: https://github.com/bitcoin-core/gui/pull/524#issuecomment-1028906108

P.s.: Thanks, @hebasto, for your help in restoring this commit.

**Comment by RandyMcMillan** [comment]
@shaavan - thanks for leaving that function in - @rebroad and I have work that utilizes this.
